### PR TITLE
Fixed overflow unsoundness in RwLock, relaxed Send/Sync impl

### PR DESCRIPTION
**Pull request by zesterer**

### CODE REVIEW

File: src/rwlock.rs [hunk 1]
```diff
@@ -248,6 +248,24 @@ impl<T: ?Sized, R: RelaxStrategy> RwLock<T, R> {
 }
 
 impl<T: ?Sized, R> RwLock<T, R> {
+    // Acquire a read lock, returning the new lock value.
+    fn acquire_reader(&self) -> usize {
+        loop {
+            let value = self.lock.load(Ordering::Acquire);
```
Comment (zakarumych) [inline src/rwlock.rs:254]: Relaxed is enough here.

**Review by zakarumych — COMMENTED** [review]
This looks significantly more expensive than single `fetch_add`.
This strategy is required when there can be multiple writers, which is not the case for `RwLock` I presume.
Otherwise it is possible to do `fetch_add` as usual, and fallback with `fetch_sub` after a threshold that should be
* large enough to be never reached practically
* small enough so any possible number of threads would not concurrently overflow the counter.

I think `usize::MAX >> 1` should do.

**Comment by zesterer** [comment]
>     * small enough so any possible number of threads would not concurrently overflow the counter.

Although I agree with this on some level, it's also just moving what was already a fairly abstract problem. There is no upper bound on how many threads could hypothetically fit into that critical section and poor scheduler behaviour is always possible. The unsoundness is not removed, just made less likely.

**Comment by zakarumych** [comment]
I personally doubt that there can be `usize::MAX / 8` threads. They would simply not fit into memory space. Unless each thread occupies 4 bytes. Which is not possible, there are more data on the stack.

**Pull request merged**